packages/firestore/test/integration/api/transactions.test.ts

```diff
@@ -63,131 +240,261 @@ apiDescribe('Database transactions', (persistence: boolean) => {
     });
   });
 
-  it('delete documents', () => {
-    return integrationHelpers.withTestDb(persistence, db => {
-      const doc = db.collection('towns').doc();
-      return doc
-        .set({ foo: 'bar' })
-        .then(() => {
-          return doc.get();
-        })
-        .then(snapshot => {
-          expect(snapshot).to.exist;
-          expect(snapshot.data()!['foo']).to.equal('bar');
-          return db.runTransaction(async transaction => {
-            transaction.delete(doc);
-          });
-        })
-        .then(() => {
-          return doc.get();
-        })
-        .then(snapshot => {
-          expect(snapshot.exists).to.equal(false);
-        });
-    });
-  });
-
-  it('get nonexistent document then create', () => {
-    return integrationHelpers.withTestDb(persistence, db => {
-      const docRef = db.collection('towns').doc();
-      return db
-        .runTransaction(transaction => {
-          return transaction.get(docRef).then(docSnap => {
-            expect(docSnap.exists).to.equal(false);
-            transaction.set(docRef, { foo: 'bar' });
-          });
-        })
-        .then(() => {
-          return docRef.get();
-        })
-        .then(snapshot => {
-          expect(snapshot.exists).to.equal(true);
-          expect(snapshot.data()).to.deep.equal({ foo: 'bar' });
-        });
-    });
-  });
-
-  it('get nonexistent document then fail patch', () => {
-    return integrationHelpers.withTestDb(persistence, db => {
-      const docRef = db.collection('towns').doc();
-      return db
-        .runTransaction(transaction => {
-          return transaction.get(docRef).then(docSnap => {
-            expect(docSnap.exists).to.equal(false);
-            transaction.update(docRef, { foo: 'bar' });
-          });
-        })
-        .then(() => expect.fail('transaction should fail'))
-        .catch((err: firestore.FirestoreError) => {
-          expect(err).to.exist;
-          expect(err.code).to.equal('invalid-argument');
-        });
+  it('run transactions after getting existing document', async () => {
+    return integrationHelpers.withTestDb(persistence, async db => {
+      const tt = new TransactionTester(db);
+      await tt.validateSequenceSuccess(
+        undefined,
+        /** docExists = */ true,
+        tt.getExisting,
```

Comment:
If you made these free functions you could avoid prefixing with `tt.` for each of these.